stdlib/LinearAlgebra/src/symmetriceigen.jl

```diff
@@ -165,25 +165,55 @@ function eigen!(A::Hermitian{T,S}, B::Hermitian{T,S}; sortby::Union{Function,Not
     GeneralizedEigen(sorteig!(vals, vecs, sortby)...)
 end
 function eigen!(A::RealHermSymComplexHerm{T,<:StridedMatrix}, B::AbstractMatrix{T}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
-    return _choleigen!(A, B, sortby)
+    return eigen!(Matrix{T}(A), B; sortby) ;
 end
 function eigen!(A::StridedMatrix{T}, B::Union{RealHermSymComplexHerm{T},Diagonal{T}}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
-    return _choleigen!(A, B, sortby)
+    return eigen!(A, Matrix{T}(B); sortby) ;
 end
-function _choleigen!(A, B, sortby)
-    U = cholesky(B).U
+
+function eigen(A::AbstractMatrix{T}, C::Cholesky{T, <:AbstractMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
+    eigen!(copy(A), copy(C); sortby)
+end
+function eigen!(A::AbstractMatrix{T}, C::Cholesky{T, <:AbstractMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
+    return _choleigen!(A, C.U; sortby)
+end
+# Cholesky decomposition based eigenvalues and eigenvectors
+function _choleigen!(A, U; sortby)
     vals, w = eigen!(UtiAUi!(A, U))
     vecs = U \ w
     GeneralizedEigen(sorteig!(vals, vecs, sortby)...)
 end
 
+function eigen(A::AbstractMatrix{T}, B::BunchKaufman{T,<:StridedMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
+    eigen!(copy(A), copy(B); sortby)
+end
+function eigen!(A::AbstractMatrix{T}, B::BunchKaufman{T,<:StridedMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
+    return _bkeigen!(A, B; sortby)
+end
+# Bunchkaufman decomposition based eigenvalues and eigenvectors
+function _bkeigen!(A, B; sortby)
+    if B.uplo == 'U'
+        vals, w = eigen!(ldiv(lu!(B.D), UiAUti!(A[B.p,B.p], B.U)); sortby)
```

Comment:
The in-place `lu!(B.D)` seems to be not working in the real symmetric case. Also, is there a specific reason this function body is not simply move to `eigen!` a few lines above? I guess the same question applies to the cholesky case above. And one more thing: this seems to be out-of-place w.r.t. `A`, right? Because `A[B.p,B.p]` allocates basically another copy of `A`. So I start to wonder what's the benefit of the whole BunchKaufman approach. Would `view`s help here instead of slicing? Presumably `UiAUti!` will be much slower on views.

Reply:
> Also, is there a specific reason this function body is not simply move to eigen! a few lines above? I guess the same question applies to the cholesky case above.

[ACCEPTED] Resolved the parts about superfluous definitions in https://github.com/aravindh-krishnamoorthy/julia/commit/0aee951c6c6d619a630dba5b07b72f39d9a1e3bb

Reply:
> The in-place lu!(B.D) seems to be not working in the real symmetric case.

[ACCEPTED] Created a "temporary" fix in https://github.com/aravindh-krishnamoorthy/julia/commit/8361afab9a23a9b57ba8503cbdf6c4d6133325d2, see the discussion in #49673 for details.

Reply:
> And one more thing: this seems to be out-of-place w.r.t. A, right? Because A[B.p,B.p] allocates basically another copy of A. So I start to wonder what's the benefit of the whole BunchKaufman approach. Would views help here instead of slicing? Presumably UiAUti! will be much slower on views.

Kindly see the discussion in #49673